lib/internal/wrap_js_stream.js

```diff
@@ -88,33 +97,56 @@ class JSStreamWrap extends Socket {
   }
 
   doShutdown(req) {
+    assert.strictEqual(this[kCurrentShutdownRequest], null);
+    this[kCurrentShutdownRequest] = req;
+
+    // TODO(addaleax): It might be nice if we could get into a state where
+    // DoShutdown() is not called on streams while a write is still pending.
+    //
+    // Currently, the only part of the code base where that happens is the
+    // TLS implementation, which calls both DoWrite() and DoShutdown() on the
+    // underlying network stream inside of its own DoShutdown() method.
+    // Working around that on the native side is not quite trivial (yet?),
+    // so for now that is supported here.
+
+    if (this[kCurrentWriteRequest] !== null)
+      return this.on('drain', () => this.doShutdown(req));
+    assert.strictEqual(this[kCurrentWriteRequest], null);
+
     const handle = this._handle;
-    const item = this._enqueue('shutdown', req);
 
     this.stream.end(() => {
       // Ensure that write was dispatched
       setImmediate(() => {
-        if (!this._dequeue(item))
-          return;
-
-        handle.finishShutdown(req, 0);
+        this.finishShutdown(handle, 0);
       });
     });
     return 0;
   }
 
+  // handle === this._handle except when called from doClose().
+  finishShutdown(handle, errCode) {
+    // The shutdown request might already have been cancelled.
+    if (this[kCurrentShutdownRequest] === null)
+      return;
+    const req = this[kCurrentShutdownRequest];
+    this[kCurrentShutdownRequest] = null;
+    handle.finishShutdown(req, errCode);
+  }
+
   doWrite(req, bufs) {
-    const self = this;
-    const handle = this._handle;
+    assert.strictEqual(this[kCurrentWriteRequest], null);
+    assert.strictEqual(this[kCurrentShutdownRequest], null);
+    this[kCurrentWriteRequest] = req;
 
-    var pending = bufs.length;
+    const handle = this._handle;
+    const self = this;
 
-    // Queue the request to be able to cancel it
-    const item = this._enqueue('write', req);
+    let pending = bufs.length;
 
     this.stream.cork();
-    for (var n = 0; n < bufs.length; n++)
-      this.stream.write(bufs[n], done);
+    for (const buf of bufs)
```

Comment:
can you use a `for(;;)` loop here? for-of is still extremely slower than `for(;;)`.